File: sdk/typescript/src/namespaces/organizations.ts
/**
 * Organizations Namespace API
 *
 * Provides a namespaced interface for organization management operations.
 * This wraps the flat client methods for a more intuitive API.
 */

import type {
  Organization,
  OrganizationMember,
  OrganizationInvitation,
  UserOrganization,
  PaginationParams,
} from '../types';

/**
 * Interface for the internal client methods used by OrganizationsAPI.
 */
interface OrganizationsClientInterface {
  request<T = unknown>(method: string, path: string, options?: Record<string, unknown>): Promise<T>;
  getOrganization(orgId: string): Promise<Organization>;
  updateOrganization(orgId: string, body: { name?: string; settings?: Record<string, unknown> }): Promise<Organization>;
  listOrganizationMembers(orgId: string, params?: PaginationParams): Promise<{ members: OrganizationMember[] }>;
  inviteToOrganization(orgId: string, body: { email: string; role?: 'admin' | 'member' }): Promise<OrganizationInvitation>;
  listOrganizationInvitations(orgId: string, params?: PaginationParams): Promise<{ invitations: OrganizationInvitation[] }>;
  revokeOrganizationInvitation(orgId: string, invitationId: string): Promise<void>;
  removeOrganizationMember(orgId: string, userId: string): Promise<void>;
  updateOrganizationMemberRole(orgId: string, userId: string, role: 'admin' | 'member'): Promise<OrganizationMember>;
  listUserOrganizations(): Promise<{ organizations: UserOrganization[] }>;
  switchOrganization(orgId: string): Promise<{ organization_id: string }>;
  setDefaultOrganization(orgId: string): Promise<{ organization_id: string }>;
  leaveOrganization(orgId: string): Promise<void>;
  listPendingInvitations(): Promise<{ invitations: OrganizationInvitation[] }>;
  acceptInvitation(token: string): Promise<{ organization_id: string; role: string }>;
}

/**
 * Organizations API namespace.
 *
 * Provides methods for organization management:
 * - Managing organization settings
 * - Member management and invitations
 * - Organization switching for users
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // Get current organization
 * const org = await client.organizations.get(orgId);
 *
 * // Invite a member
 * await client.organizations.invite(orgId, {
 *   email: 'newuser@example.com',
 *   role: 'member',
 * });
 *
 * // List user's organizations
 * const { organizations } = await client.organizations.listMine();
 *
 * // Switch active organization
 * await client.organizations.switch(orgId);
 * ```
 */
export class OrganizationsAPI {
  constructor(private client: OrganizationsClientInterface) {}

  /**
   * Get an organization by ID.
   */
  async get(orgId: string): Promise<Organization> {
    return this.client.getOrganization(orgId);
  }

  /**
   * Update an organization.
   */
  async update(
    orgId: string,
    body: { name?: string; settings?: Record<string, unknown> }
  ): Promise<Organization> {
    return this.client.updateOrganization(orgId, body);
  }

  /**
   * List organization members.
   */
  async listMembers(
    orgId: string,
    params?: PaginationParams
  ): Promise<{ members: OrganizationMember[] }> {
    return this.client.listOrganizationMembers(orgId, params);
  }

  /**
   * Invite a user to the organization.
   */
  async invite(
    orgId: string,
    body: { email: string; role?: 'admin' | 'member' }
  ): Promise<OrganizationInvitation> {
    return this.client.inviteToOrganization(orgId, body);
  }

  /**
   * List pending invitations for an organization.
   */
  async listInvitations(
    orgId: string,
    params?: PaginationParams
  ): Promise<{ invitations: OrganizationInvitation[] }> {
    return this.client.listOrganizationInvitations(orgId, params);
  }

  /**
   * Revoke an invitation.
   */
  async revokeInvitation(orgId: string, invitationId: string): Promise<void> {
    return this.client.revokeOrganizationInvitation(orgId, invitationId);
  }

  /**
   * Remove a member from the organization.
   */
  async removeMember(orgId: string, userId: string): Promise<void> {
    return this.client.removeOrganizationMember(orgId, userId);
  }

  /**
   * Update a member's role.
   */
  async updateMemberRole(
    orgId: string,
    userId: string,
    role: 'admin' | 'member'
  ): Promise<OrganizationMember> {
    return this.client.updateOrganizationMemberRole(orgId, userId, role);
  }

  /**
   * List the current user's organizations.
   */
  async listMine(): Promise<{ organizations: UserOrganization[] }> {
    return this.client.listUserOrganizations();
  }

  /**
   * Switch to a different organization.
   */
  async switch(orgId: string): Promise<{ organization_id: string }> {
    return this.client.switchOrganization(orgId);
  }

  /**
   * Set the default organization for the current user.
   */
  async setDefault(orgId: string): Promise<{ organization_id: string }> {
    return this.client.setDefaultOrganization(orgId);
  }

  /**
   * Leave an organization.
   */
  async leave(orgId: string): Promise<void> {
    return this.client.leaveOrganization(orgId);
  }

  /**
   * List pending invitations for the current user.
   */
  async listMyInvitations(): Promise<{ invitations: OrganizationInvitation[] }> {
    return this.client.listPendingInvitations();
  }

  /**
   * Accept an organization invitation.
   */
  async acceptInvitation(token: string): Promise<{ organization_id: string; role: string }> {
    return this.client.acceptInvitation(token);
  }

  /**
   * Invite member to org.
   */
  async inviteMember(orgId: string): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/org/${orgId}/invite`) as Promise<Record<string, unknown>>;
  }

  /**
   * Get pending invitations.
   */
  async getPendingInvitations(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/invitations/pending', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * List user organizations.
   */
  async listUserOrganizations(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/user/organizations', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get default organization.
   */
  async getDefaultOrganization(params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/user/organizations/default', { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Switch active organization.
   */
  async switchOrganization(data?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/user/organizations/switch', { json: data }) as Promise<Record<string, unknown>>;
  }

  /**
   * List organization invitations.
   */
  async getOrgInvitations(orgId: string, params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/org/${orgId}/invitations`, { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Get organization members.
   */
  async getOrgMembers(orgId: string, params?: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/org/${orgId}/members`, { params }) as Promise<Record<string, unknown>>;
  }

  /**
   * Update a member's role in an organization.
   */
  async setMemberRole(orgId: string, memberId: string, data: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/org/${orgId}/members/${memberId}/role`, { json: data }) as Promise<Record<string, unknown>>;
  }

  /**
   * Accept an invitation.
   */
  async acceptInvite(invitationId: string): Promise<Record<string, unknown>> {
    return this.client.request('GET', `/api/v1/invitations/${invitationId}/accept`) as Promise<Record<string, unknown>>;
  }

  /**
   * Get an organization by ID (request-based).
   *
   * @param orgId - Organization ID
   */
  async getOrg(orgId: string): Promise<Organization> {
    return this.client.request('GET', `/api/v1/org/${orgId}`);
  }

  /**
   * Remove a member from an organization (request-based).
   *
   * @param orgId - Organization ID
   * @param userId - User ID to remove
   */
  async removeMemberDirect(orgId: string, userId: string): Promise<void> {
    return this.client.request('GET', `/api/v1/org/${orgId}/members/${userId}`);
  }

  /**
   * Revoke an invitation (request-based).
   *
   * @param orgId - Organization ID
   * @param invitationId - Invitation ID to revoke
   */
  async revokeInvitationDirect(orgId: string, invitationId: string): Promise<void> {
    return this.client.request('GET', `/api/v1/org/${orgId}/invitations/${invitationId}`);
  }

  /**
   * List all organizations.
   *
   * @route GET /api/v1/org
   */
  async listOrgs(params?: { limit?: number; offset?: number }): Promise<{ organizations: Organization[]; total: number }> {
    return this.client.request('GET', '/api/v1/org', { params: params as Record<string, unknown> });
  }

  /**
   * List all tenants.
   *
   * @route GET /api/v1/tenants
   */
  async listTenants(params?: { limit?: number; offset?: number }): Promise<Record<string, unknown>> {
    return this.client.request('GET', '/api/v1/tenants', { params: params as Record<string, unknown> });
  }

}

File: sdk/python/aragora_sdk/namespaces/organizations.py
"""
Organizations Namespace API

Provides methods for organization management, member management, and invitations.

Features:
- Organization details and settings
- Member listing and role management
- Team invitations
- Multi-organization support
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class OrganizationsAPI:
    """
    Synchronous Organizations API.

    Provides methods for organization management:
    - Get and update organization details
    - List and manage organization members
    - Send and manage invitations
    - Switch between organizations

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> org = client.organizations.get("org-123")
        >>> members = client.organizations.list_members("org-123")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Organization Details
    # ===========================================================================

    def list_orgs(self) -> dict[str, Any]:
        """
        List all organizations.

        Returns:
            Dict with organizations array.
        """
        return self._client.request("GET", "/api/v1/org")

    def get(self, org_id: str) -> dict[str, Any]:
        """
        Get organization details.

        Args:
            org_id: Organization ID

        Returns:
            Dict with organization info including name, slug, tier, member count
        """
        return self._client.request("GET", f"/api/v1/org/{org_id}")

    def update(
        self,
        org_id: str,
        name: str | None = None,
        settings: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """
        Update organization settings.

        Args:
            org_id: Organization ID
            name: New organization name
            settings: Settings dict to update

        Returns:
            Updated organization info
        """
        data: dict[str, Any] = {}
        if name is not None:
            data["name"] = name
        if settings is not None:
            data["settings"] = settings

        return self._client.request("PUT", f"/api/v1/org/{org_id}", json=data)

    # ===========================================================================
    # User Organizations
    # ===========================================================================

    def list_user_organizations(self) -> dict[str, Any]:
        """
        List organizations the current user belongs to.

        Returns:
            Dict with organizations array
        """
        return self._client.request("GET", "/api/v1/user/organizations")

    def switch_organization(self, org_id: str) -> dict[str, Any]:
        """
        Switch the active organization for the current user.

        Args:
            org_id: Organization ID to switch to

        Returns:
            Dict with organization info
        """
        return self._client.request(
            "POST", "/api/v1/user/organizations/switch", json={"org_id": org_id}
        )

    def set_default_organization(self, org_id: str) -> dict[str, Any]:
        """
        Set the default organization for the current user.

        Args:
            org_id: Organization ID to set as default

        Returns:
            Dict with success status
        """
        return self._client.request(
            "POST", "/api/v1/user/organizations/default", json={"org_id": org_id}
        )

    def list_members(self, org_id: str) -> dict[str, Any]:
        """
        List organization members.

        Args:
            org_id: Organization ID

        Returns:
            Dict with members array and count
        """
        return self._client.request("GET", f"/api/v1/org/{org_id}/members")

    def remove_member(self, org_id: str, user_id: str) -> dict[str, Any]:
        """
        Remove a member from the organization.

        Args:
            org_id: Organization ID
            user_id: User ID to remove

        Returns:
            Dict with success message
        """
        return self._client.request("DELETE", f"/api/v1/org/{org_id}/members/{user_id}")

    def update_member_role(self, org_id: str, user_id: str, role: str) -> dict[str, Any]:
        """
        Update a member's role in the organization.

        Args:
            org_id: Organization ID
            user_id: User ID to update
            role: New role ('member' or 'admin')

        Returns:
            Dict with updated role info
        """
        return self._client.request(
            "PUT", f"/api/v1/org/{org_id}/members/{user_id}/role", json={"role": role}
        )

    # ===========================================================================
    # Invitations
    # ===========================================================================

    def invite_member(self, org_id: str, email: str, role: str = "member") -> dict[str, Any]:
        """
        Invite a user to the organization.

        Args:
            org_id: Organization ID
            email: Email address to invite
            role: Role for the new member ('member' or 'admin')

        Returns:
            Dict with invitation info including invite link
        """
        return self._client.request(
            "POST", f"/api/v1/org/{org_id}/invite", json={"email": email, "role": role}
        )

    def list_invitations(self, org_id: str) -> dict[str, Any]:
        """
        List pending invitations for an organization.

        Args:
            org_id: Organization ID

        Returns:
            Dict with invitations array and counts
        """
        return self._client.request("GET", f"/api/v1/org/{org_id}/invitations")

    def revoke_invitation(self, org_id: str, invitation_id: str) -> dict[str, Any]:
        """
        Revoke a pending invitation.

        Args:
            org_id: Organization ID
            invitation_id: Invitation ID to revoke

        Returns:
            Dict with success message
        """
        return self._client.request("DELETE", f"/api/v1/org/{org_id}/invitations/{invitation_id}")

    def get_pending_invitations(self) -> dict[str, Any]:
        """
        Get pending invitations for the current user.

        Returns:
            Dict with invitations array
        """
        return self._client.request("GET", "/api/v1/invitations/pending")

    def accept_invitation(self, token: str) -> dict[str, Any]:
        """
        Accept an invitation to join an organization.

        Args:
            token: Invitation token

        Returns:
            Dict with organization info and role
        """
        return self._client.request("POST", f"/api/v1/invitations/{token}/accept")

    # ===========================================================================
    # Tenants
    # ===========================================================================

    def list_tenants(self) -> dict[str, Any]:
        """
        List all tenants.

        Returns:
            Dict with tenants array.
        """
        return self._client.request("GET", "/api/v1/tenants")


class AsyncOrganizationsAPI:
    """
    Asynchronous Organizations API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     org = await client.organizations.get("org-123")
        ...     members = await client.organizations.list_members("org-123")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Organization Details
    # ===========================================================================

    async def list_orgs(self) -> dict[str, Any]:
        """List all organizations."""
        return await self._client.request("GET", "/api/v1/org")

    async def get(self, org_id: str) -> dict[str, Any]:
        """Get organization details."""
        return await self._client.request("GET", f"/api/v1/org/{org_id}")

    async def update(
        self,
        org_id: str,
        name: str | None = None,
        settings: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        """Update organization settings."""
        data: dict[str, Any] = {}
        if name is not None:
            data["name"] = name
        if settings is not None:
            data["settings"] = settings

        return await self._client.request("PUT", f"/api/v1/org/{org_id}", json=data)

    # ===========================================================================
    # User Organizations
    # ===========================================================================

    async def list_user_organizations(self) -> dict[str, Any]:
        """List organizations the current user belongs to."""
        return await self._client.request("GET", "/api/v1/user/organizations")

    async def switch_organization(self, org_id: str) -> dict[str, Any]:
        """Switch the active organization for the current user."""
        return await self._client.request(
            "POST", "/api/v1/user/organizations/switch", json={"org_id": org_id}
        )

    async def set_default_organization(self, org_id: str) -> dict[str, Any]:
        """Set the default organization for the current user."""
        return await self._client.request(
            "POST", "/api/v1/user/organizations/default", json={"org_id": org_id}
        )

    async def list_members(self, org_id: str) -> dict[str, Any]:
        """List organization members."""
        return await self._client.request("GET", f"/api/v1/org/{org_id}/members")

    async def remove_member(self, org_id: str, user_id: str) -> dict[str, Any]:
        """Remove a member from the organization."""
        return await self._client.request("DELETE", f"/api/v1/org/{org_id}/members/{user_id}")

    async def update_member_role(self, org_id: str, user_id: str, role: str) -> dict[str, Any]:
        """Update a member's role in the organization."""
        return await self._client.request(
            "PUT", f"/api/v1/org/{org_id}/members/{user_id}/role", json={"role": role}
        )

    # ===========================================================================
    # Invitations
    # ===========================================================================

    async def invite_member(self, org_id: str, email: str, role: str = "member") -> dict[str, Any]:
        """Invite a user to the organization."""
        return await self._client.request(
            "POST", f"/api/v1/org/{org_id}/invite", json={"email": email, "role": role}
        )

    async def list_invitations(self, org_id: str) -> dict[str, Any]:
        """List pending invitations for an organization."""
        return await self._client.request("GET", f"/api/v1/org/{org_id}/invitations")

    async def revoke_invitation(self, org_id: str, invitation_id: str) -> dict[str, Any]:
        """Revoke a pending invitation."""
        return await self._client.request(
            "DELETE", f"/api/v1/org/{org_id}/invitations/{invitation_id}"
        )

    async def get_pending_invitations(self) -> dict[str, Any]:
        """Get pending invitations for the current user."""
        return await self._client.request("GET", "/api/v1/invitations/pending")

    async def accept_invitation(self, token: str) -> dict[str, Any]:
        """Accept an invitation to join an organization."""
        return await self._client.request("POST", f"/api/v1/invitations/{token}/accept")

    # Tenants
    async def list_tenants(self) -> dict[str, Any]:
        """List all tenants."""
        return await self._client.request("GET", "/api/v1/tenants")
